src/ray/core_worker/actor_manager.cc

```diff
@@ -180,4 +194,67 @@ std::vector<ObjectID> ActorManager::GetActorHandleIDsFromHandles() {
   return actor_handle_ids;
 }
 
+void ActorManager::ResolveActorsLocations() {
+  absl::MutexLock lock(&mutex_);
+  for (auto it = actors_pending_location_resolution_.begin();
+       it != actors_pending_location_resolution_.end();) {
+    auto current = it++;
+    const auto &actor_id = *current;
+
+    const std::unique_ptr<ActorHandle> &actor_handle = GetActorHandleInternal(actor_id);
+    const ClientID &node_id =
+        ClientID::FromBinary(actor_handle->GetOwnerAddress().raylet_id());
+
+    if (actor_handle->IsPersistedToGCS()) {
+      // if an actor is already persisted to GCS, GCS will be responsible for lifecycle of
+      // actors, so we don't need to track them anymore.
+      RAY_CHECK(current != actors_pending_location_resolution_.end());
+      actors_pending_location_resolution_.erase(current);
+    } else {
+      // https://github.com/ray-project/ray/pull/8679/files
+      // Run a protocol to resolve actors location that haven't been registered to GCS.
+      RAY_CHECK_OK(gcs_client_->Workers().AsyncGet(
+          WorkerID::FromBinary(actor_handle->GetOwnerAddress().worker_id()),
+          [this, actor_id, node_id](Status status,
+                                    const boost::optional<rpc::WorkerTableData> &result) {
+            if (!status.ok()) {
+              return;
+            }
+
+            bool worker_or_node_failed = false;
+            if (result && !result->is_alive()) {
+              worker_or_node_failed = true;
+            } else {
+              // Check node failure. We should do this because worker failure event is not
+              // reported when nodes fail.
+              const auto &optional_node_info = gcs_client_->Nodes().Get(node_id);
+              RAY_CHECK(optional_node_info)
+                  << "Node information for an actor_id, " << actor_id << " is not found.";
+              if (optional_node_info->state() ==
+                  rpc::GcsNodeInfo_GcsNodeState::GcsNodeInfo_GcsNodeState_DEAD) {
+                worker_or_node_failed = true;
+              }
+            }
+
+            if (worker_or_node_failed) {
+              // We should make sure one more time that an actor is not registered to GCS
+              // before resolving actor's location to avoid race condition.
+              RAY_CHECK_OK(gcs_client_->Actors().AsyncGet(
+                  actor_id,
+                  [this, actor_id](Status status,
```

Comment:
I know it's a bit annoying with the async code, but it'd be nice to pull out some of this code so that it's not so deeply nested.

Reply:
Will take out the callback.

Reply:
I took out the whole function instead.